Mage.Sets/src/mage/cards/k/KaghaShadowArchdruid.java

```diff
@@ -0,0 +1,154 @@
+        super(ownerId, setInfo, new CardType[]{CardType.CREATURE}, "{2}{B}{G}");
+        
+        this.supertype.add(SuperType.LEGENDARY);
+        this.subtype.add(SubType.ELF);
+        this.subtype.add(SubType.DRUID);
+        this.power = new MageInt(1);
+        this.toughness = new MageInt(4);
+
+        // Whenever Kagha, Shadow Archdruid attacks, it gains deathtouch until end of turn. Mill two cards.
+        Ability ability = new AttacksTriggeredAbility(new GainAbilitySourceEffect(
+            DeathtouchAbility.getInstance(),
+            Duration.EndOfTurn));
+        ability.addEffect(new MillCardsControllerEffect(2));
+        this.addAbility(ability);
+
+        // Once during each of your turns, you may play a land or cast a permanent spell from among cards in your graveyard that were put there from your library this turn.
+        this.addAbility(new SimpleStaticAbility(Zone.BATTLEFIELD, new KaghaShadowArchdruidEffect())
+            .setIdentifier(MageIdentifier.KaghaShadowArchdruidWatcher),
+            new KaghaShadowArchdruidWatcher());
+    }
+
+    private KaghaShadowArchdruid(final KaghaShadowArchdruid card) {
+        super(card);
+    }
+
+    @Override
+    public KaghaShadowArchdruid copy() {
+        return new KaghaShadowArchdruid(this);
+    }
+}
+
+class KaghaShadowArchdruidEffect extends AsThoughEffectImpl {
+
+    KaghaShadowArchdruidEffect() {
+        super(AsThoughEffectType.PLAY_FROM_NOT_OWN_HAND_ZONE, Duration.WhileOnBattlefield, Outcome.Benefit, true);
+        this.staticText = "Once during each of your turns, you may play a land or cast a permanent spell from among cards in your graveyard that were put there from your library this turn.";
+    }
+
+    private KaghaShadowArchdruidEffect(final KaghaShadowArchdruidEffect effect) {
+        super(effect);
+    }
+
+    @Override
+    public KaghaShadowArchdruidEffect copy() {
+        return new KaghaShadowArchdruidEffect(this);
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        return true;
+    }
+
+    @Override
+    public boolean applies(UUID objectId, Ability source, UUID affectedControllerId, Game game) {
+        if (source.isControlledBy(affectedControllerId)
+                && Zone.GRAVEYARD.equals(game.getState().getZone(objectId))
+                && game.isActivePlayer(affectedControllerId)) {
+            Card card = game.getCard(objectId);
+            Permanent sourceObject = source.getSourcePermanentIfItStillExists(game);
+            KaghaShadowArchdruidWatcher watcher = game.getState().getWatcher(KaghaShadowArchdruidWatcher.class);
+            if (card != null && sourceObject != null 
+                    && card.isOwnedBy(affectedControllerId)
+                    && watcher != null
+                    && watcher.checkCard(card, game)) {
+                return watcher.abilityNotUsed(new MageObjectReference(sourceObject, game));
+            }
```

Comment:
Can be a big return once the two Set from the Watcher are package-private instead of private:

```java
return card != null 
    && sourceObject != null 
    && card.isOwnedBy(affectedControllerId)
    && watcher != null
    && card.isPermanent(game)
    && watcher.morSet
              .stream()
              .anyMatch(mor -> mor.refersTo(card, game));
    && watcher.usedFor.contains(new MageObjectReference(sourceObject, game));

```

I think that one the intent behind xenohedron's comment.

Reply:
Yeah, one big return, but should still use abilityNotUsed() and checkCard() methods from the watcher with fields private.